yarn-project/ethereum/src/gas_utils.ts

```diff
@@ -0,0 +1,216 @@
+
+export interface GasConfig {
+  /**
+   * How much to increase gas price by each attempt (percentage)
+   */
+  bufferPercentage: bigint;
+  /**
+   * Maximum gas price in gwei
+   */
+  maxGwei: bigint;
+  /**
+   * Minimum gas price in gwei
+   */
+  minGwei: bigint;
+  /**
+   * Priority fee in gwei
+   */
+  priorityFeeGwei: bigint;
+}
+
+export interface TransactionMonitorConfig {
+  /**
+   * Maximum number of speed-up attempts
+   */
+  maxAttempts: number;
+  /**
+   * How often to check tx status
+   */
+  checkIntervalMs: number;
+  /**
+   * How long before considering tx stalled
+   */
+  stallTimeMs: number;
+  /**
+   * How much to increase gas price by each attempt (percentage)
+   */
+  gasPriceIncrease: bigint;
+}
+
+const DEFAULT_GAS_CONFIG: GasConfig = {
+  bufferPercentage: 20n,
+  maxGwei: 500n,
+  minGwei: 1n,
+  priorityFeeGwei: 2n,
+};
+
+const DEFAULT_MONITOR_CONFIG: Required<TransactionMonitorConfig> = {
+  maxAttempts: 3,
+  checkIntervalMs: 30_000,
+  stallTimeMs: 180_000,
+  gasPriceIncrease: 50n,
+};
+
+export class GasUtils {
+  private readonly gasConfig: GasConfig;
+  private readonly monitorConfig: TransactionMonitorConfig;
+
+  constructor(
+    private readonly publicClient: PublicClient,
+    private readonly logger?: DebugLogger,
+    gasConfig?: GasConfig,
+    monitorConfig?: TransactionMonitorConfig,
+  ) {
+    this.gasConfig! = gasConfig ?? DEFAULT_GAS_CONFIG;
+    this.monitorConfig! = monitorConfig ?? DEFAULT_MONITOR_CONFIG;
+  }
+
+  /**
+   * Gets the current gas price with safety buffer and bounds checking
+   */
+  public async getGasPrice(): Promise<bigint> {
+    const block = await this.publicClient.getBlock({ blockTag: 'latest' });
+    const baseFee = block.baseFeePerGas ?? 0n;
+    const priorityFee = this.gasConfig.priorityFeeGwei * WEI_CONST;
+
+    const baseWithBuffer = baseFee + (baseFee * this.gasConfig.bufferPercentage) / 100n;
+    const totalGasPrice = baseWithBuffer + priorityFee;
+
+    const maxGasPrice = this.gasConfig.maxGwei * WEI_CONST;
+    const minGasPrice = this.gasConfig.minGwei * WEI_CONST;
+
+    let finalGasPrice: bigint;
+    if (totalGasPrice < minGasPrice) {
+      finalGasPrice = minGasPrice;
+    } else if (totalGasPrice > maxGasPrice) {
+      finalGasPrice = maxGasPrice;
+    } else {
+      finalGasPrice = totalGasPrice;
+    }
+
+    this.logger?.debug(
+      `Gas price calculation: baseFee=${baseFee}, withBuffer=${baseWithBuffer}, priority=${priorityFee}, final=${finalGasPrice}`,
+    );
+
+    return finalGasPrice;
+  }
+
+  /**
+   * Estimates gas with retries and adds buffer
+   */
+  public estimateGas<T extends (...args: any[]) => Promise<bigint>>(estimator: T): Promise<bigint> {
+    return retry(
+      async () => {
+        const gas = await estimator();
+        return gas + (gas * this.gasConfig.bufferPercentage) / 100n;
+      },
+      'gas estimation',
+      makeBackoff([1, 2, 3]),
+      this.logger,
+      false,
+    );
```

Comment:
Why the `retry` here as opposed to on other rpc calls? If we are seeing flaky rpc endpoints where we need to retry, we should tackle that at a lower level (on viem's "transport" most likely).